### ISSUE TITLE
Add a .hasNext() assertion method for Iterator<?>

### ISSUE THREAD

**Issue by fge** [joel-costigliola/assertj-core#414]
I'd have thought that this method already existed but it appears that in fact, no...

It'd be a nice-to-have; along with `.doesNotHaveNext()`.

Comments?

**Comment by joel-costigliola** [comment]
I would prefer `isExhausted`

**Comment by fge** [comment]
Yes indeed, that is a better name (or `.hasNoMoreElements()`; bah)

**Comment by AlexBischof** [comment]
Why not use *isNotEmpty*? It does internally check for *hasNext()* and should semantically fit, should it?

Although the behavior is a little bit strange to me in the following code
``` java
        ArrayList arrayList = new ArrayList();
        arrayList.add("Luke");
        arrayList.add( "Leia");
        Iterator<String> it = arrayList.iterator();
        while (true) {
            assertThat(it).isNotEmpty();
            String next = it.next();
            System.out.println(next);
        }
```
*java.util.NoSuchElementException*

**Comment by joel-costigliola** [comment]
Your `while` condition is wrong, it should `it.hasNext()`.

I think `exhausted` is the usual term for an `Iterator` that don't have any elements left.

**Comment by fge** [comment]
I personally agree that the code extract of @AlexBischof should work.

Note that I haven't looked at the code of assertj in detail yet, but it looks to me that `.isNotEmpty()` just does something like this (and yes, I am quite sure that this is not how things work internally):

```java
// "it" is the iterator here
public boolean isNotEmpty()
{
    try {
        underlyingIterator.next();
        return true;
    } catch (NoSuchElementException ignored) {
        return false;
    }
}
```

which in turn would trigger the `NoSuchElementException` in the code example.

Now, another important thing to bear in mind here is that an `Iterable` and and `Iterator` are not quite the same thing. An `Iterable` is supposed to produce a fresh `Iterator` each time you call its `.iterator()` method; but this is not quite the same thing as testing an `Iterator` in itself.

In other words, an `Iterator` produced by an `Iterable` is independent from any other `Iterator` produced by that same `Iterable`.

As such I believe that it is incorrect to "promote" the test of an `Iterator` to that of `Iterable`, and that dedicated assertions for `Iterator`s are worth considering.

For instance:

```java
assertThat(iterator).hasNext();
assertThat(iterator).nextElement().is*();
// or even chain
```

Comments?

**Comment by joel-costigliola** [comment]
[Javadoc](http://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/Assertions.html#assertThat%28java.util.Iterator%29) states clearly that assertions on an Iterator will consume it in most of the case. It was an easy solution to provide `Iterator` assertions.

Having said that, I agree that checking that an Iterator has elements left should not consume it.  

@fge My gut feeling is that having specific `Iterator` assertions will require a lot of work if you want to support the richness of `Iterable` assertions.

**Comment by AlexBischof** [comment]
+1 for a specific assertion because it would be a much cleaner integration.

**Comment by stovocor** [comment]
I am interested in working on this. Is the initial analysis still valid after three years?

**Comment by joel-costigliola** [comment]
I think that @fge is right, the best solution would be to have `Iterator` assertions that only deals with the `Iterator` API including `isExhausted()`, `hasNext()`, ...

Navigational assertions like `assertThat(iterator).nextElement().isEqualto(...);` would be nice to have but are tricky to implement because of soft assertions proxying, I'd rather create a separate issue for them.

**Comment by stovocor** [comment]
Yes, I would like to work on that.

It is very likely that there will be questions regarding the `toterable(iterator)` and `asIterable()` methods, but I will ask them here when I get there.

**Comment by stovocor** [comment]
Okay, first question for my initial commit. I have some issues satisfying the generic constraints for the tests, especially the unchecked warning in `AbstractIteratorAssert`. Do I need something like `ConcreteIterableAssert` as an intermediate type?

**Comment by joel-costigliola** [comment]
@stovocor can you create a PR ? It makes it easier to review changes, thanks!

**Comment by stovocor** [comment]
> can you create a PR ? It makes it easier to review changes, thanks!

I do not think that the changes are ready to be merged, but I created the PR as you asked.

**Comment by joel-costigliola** [comment]
that's alright, it just to have a look at it and give you some feedback for the next steps

**Comment by stovocor** [comment]
@joel-costigliola I just rebased this against the current master. Can you please tell me how I should continue with this pull request?

**Comment by joel-costigliola** [comment]
@stovocor I'll have alook this we and let you know (remind me if I fail to!)
